File: ruby/bob/bob.rb
class Bob
  def hey(remark)
    if is_silent?(remark)
      'Fine. Be that way!'
    elsif is_yelling?(remark)
      'Whoa, chill out!'
    elsif is_question?(remark)
      'Sure.'
    else
      'Whatever.'
    end
  end

  private

    def is_silent?(remark)
      remark.strip == ''
    end

    def is_yelling?(remark)
      remark.upcase == remark && remark.downcase != remark
    end

    def is_question?(remark)
      remark[-1] == '?'
    end
end

File: java/bob/src/main/java/Bob.java
public class Bob {

  private static final String insouciantResponse = "Whoa, chill out!";

  private static final String lackadaisicalAnswer = "Sure.";

  private static final String petulantResponse = "Fine. Be that way!";

  private static final String desultoryResponse = "Whatever.";

  public final String hey(String goesInEar) {
    if (hasExcitement(goesInEar)) return insouciantResponse;
    if (isQuestion(goesInEar)) return lackadaisicalAnswer;
    if (isSilent(goesInEar)) return petulantResponse;

    return desultoryResponse;
  }

  private final boolean hasExcitement(String goesInEar) {
    return goesInEar.chars().anyMatch(Character::isLetter)
      && goesInEar.equals(goesInEar.toUpperCase());
  }

  private final boolean isSilent(String goesInEar) {
    return goesInEar.trim().isEmpty();
  }

  private final boolean isQuestion(String goesInEar) {
    return goesInEar.endsWith("?");
  }
}
